File: spiral.c
#include <stdio.h>
int main(void) {
	// your code goes here
	int row, col;
	int i, j;
	int t,b,l,r;
	printf("Enter rows\n");
	scanf("%d", &row);
	printf("enter cols\n");
	scanf("%d", &col);
	printf("rows : %d and cols : %d\n", row,col);
	int s[row][col];
	printf("enter array elems\n");
	for(i=0; i<row; i++)
	{
		printf("row %d\n", i);
		for(j=0; j<col; j++)
		{
			scanf("%d", &s[i][j]);
		}
	}
	for(i=0; i<row; i++)
	{
		printf("row %d\n", i);
		for(j=0; j<col; j++)
		{
			printf("%d", s[i][j]);
		}
		printf("\n");
	}
	t=0;
	b= row-1;
	l=0;
	r=col-1;
	int dir = 0;
	printf("l: %d r: %d, t: %d, b: %d\n", l, r, t, b);
	int x, y;
	printf("\n");
	while(l<=r && t<=b)
	{
		if(dir==0)
		{
			printf("left to right\n");
			for(x=l; x<=r; x++)
			{
				printf("%d", s[t][x]);
			}
			t++;
			printf("l: %d r: %d, t: %d, b: %d\n", l, r, t, b);
		}
		else if(dir==1)
		{
			printf("top to bottom\n");
			for(y=t; y<=b; y++)
			{
				printf("%d", s[y][r]);
			}
			r--;
			printf("l: %d r: %d, t: %d, b: %d\n", l, r, t, b);
		}
		else if(dir==2)
		{
			printf("right to left\n");
			for(x=r; x>=l; x--)
			{
				printf("%d", s[b][x]);
			}
			b--;
			printf("l: %d r: %d, t: %d, b: %d\n", l, r, t, b);
		}
		else if(dir==3)
		{
			printf("bottom to top\n");
			for(y=b;y>=t; y--)
			{
				printf("%d", s[y][l]);
			}
			l++;
			printf("l: %d r: %d, t: %d, b: %d\n", l, r, t, b);
		}
		printf("\n");
		dir = (dir+1)%4;
	}
	
}

File: interview_probs/spiral.py
def printspiral(A, m, n):

    print "rows -->" , m
    print "columns -->", n
    print "2d array --> \n"
    for x in xrange(m):
        print A[x]
    t = 0
    b = m-1
    l=0
    r=n-1
    dir = 0
    ret = []
    while (l<=r and t<=b):
        if dir == 0:
            print "left to right"
            for i in xrange(l,r+1):
                print A[t][i]
                ret.append(A[t][i])
            t=t+1
            print ""
        elif dir == 1:
            print "top to bottom"
            for i in xrange(t,b+1):
                print A[i][r]
                ret.append(A[i][r])
            r=r-1
            print ""
        elif dir == 2:
            print "right to left"
            for i in xrange(r,l-1,-1):
                print A[b][i]
                ret.append(A[b][i])
            b=b-1
            print ""
        elif dir == 3:
            print "bottom to top"
            for i in xrange(b,t-1,-1):
                print A[i][l]
                ret.append(A[i][l])
            l=l+1
            print ""
        #change direction
        dir = (dir+1)%4

    return ret




# longer array for testing
#m = 4; n = 4
#A = [[1,2,3,4],[5,6,7,8],[9,10,11,12],[13,14,15,16]]
m = 3; n = 3
A=[[1,2,3],[4,5,6],[7,8,9]]
#m=4; n=2
#A = [[1,2],[3,4],[5,6],[7,8]]
#m = 2; n= 3
#A = [[1,2,3],[4,5,6]]
x = printspiral(A, m, n)
print x
